src/statement-list/statement_list_typecheck.cpp

```diff
@@ -30,87 +42,1079 @@ bool statement_list_typecheck(
+  fc_bit = false;
+  or_bit = false;
+  rlo_bit = false_exprt();
+}
+
+void statement_list_typecheckt::typecheck_statement_list_set(
+  const codet &op_code,
+  symbolt &tia_element)
+{
+  typecheck_instruction_with_operand(op_code);
+  const irep_idt &identifier = to_symbol_expr(op_code.op0()).get_identifier();
+  const exprt lhs = typecheck_identifier(tia_element, identifier);
+  const code_assignt assignment{lhs, true_exprt()};
+  const code_ifthenelset ifthen{rlo_bit, assignment};
+  tia_element.value.add_to_operands(ifthen);
+  fc_bit = false;
+  or_bit = false;
+}
+
+void statement_list_typecheckt::typecheck_statement_list_reset(
+  const codet &op_code,
+  symbolt &tia_element)
+{
+  typecheck_instruction_with_operand(op_code);
+  const irep_idt &identifier = to_symbol_expr(op_code.op0()).get_identifier();
+  const exprt lhs = typecheck_identifier(tia_element, identifier);
+  const code_assignt assignment{lhs, false_exprt()};
+  const code_ifthenelset ifthen{rlo_bit, assignment};
+  tia_element.value.add_to_operands(ifthen);
+  fc_bit = false;
+  or_bit = false;
+}
+
+void statement_list_typecheckt::typecheck_statement_list_call(
+  const codet &op_code,
+  symbolt &tia_element)
+{
+  typecheck_instruction_with_operand(op_code);
+  const symbol_exprt &call_operand = to_symbol_expr(op_code.op0());
+  if(symbol_table.has_symbol(call_operand.get_identifier()))
+    typecheck_called_tia_element(op_code, tia_element);
+  else if(call_operand.get_identifier() == CPROVER_ASSUME)
+    typecheck_CPROVER_assume(op_code, tia_element);
+  else if(call_operand.get_identifier() == CPROVER_ASSERT)
+    typecheck_CPROVER_assert(op_code, tia_element);
+  else
+  {
+    error() << "Called function could not be found" << eom;
+    throw TYPECHECK_ERROR;
+  }
+  fc_bit = false;
+  or_bit = false;
+}
+
+void statement_list_typecheckt::typecheck_statement_list_accu_int_arith(
+  const codet &op_code)
+{
+  typecheck_instruction_without_operand(op_code);
+  if(accumulator.size() < 2)
+  {
+    error() << "Not enough operands in the accumulator" << eom;
+    throw TYPECHECK_ERROR;
+  }
+  const exprt &accu1 = accumulator.back();
+  const exprt &accu2 = accumulator.at(accumulator.size() - 2);
+
+  // Are both operands integers?
+  if(!(can_cast_type<signedbv_typet>(accu1.type()) &&
+       can_cast_type<signedbv_typet>(accu2.type())))
+  {
+    error() << "Operands of integer addition are no integers" << eom;
+    throw TYPECHECK_ERROR;
+  }
+}
+
+void statement_list_typecheckt::typecheck_statement_list_accu_dint_arith(
+  const codet &op_code)
+{
+  typecheck_instruction_without_operand(op_code);
+  if(accumulator.size() < 2)
+  {
+    error() << "Not enough operands in the accumulator" << eom;
+    throw TYPECHECK_ERROR;
+  }
+  const exprt &accu1 = accumulator.back();
+  const exprt &accu2 = accumulator.at(accumulator.size() - 2);
+
+  // Are both operands double integers?
+  if(!(can_cast_type<signedbv_typet>(accu1.type()) &&
```

Comment:
Technically, this only checks if it is an `INT` or a `DINT`. Check for `get_width` as well.